### ISSUE TITLE
Extend 'Simplify interpolation' to add FormattableString.Invariant

### ISSUE THREAD

**Issue by jnm2** [dotnet/roslyn#52581]
The result of following https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1310 or https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1307 is often that you end up with `.ToString(CultureInfo.InvariantCulture)` in your code. In https://github.com/dotnet/roslyn/pull/52564, I extended 'Simplify interpolation' to recognize that this simplification is possible without changing meaning:

```cs
FormattableString.Invariant($"{first.ToString("0000", CultureInfo.InvariantCulture)}{second.ToString("0000", CultureInfo.InvariantCulture)}")

// 💡 Simplify interpolation:
FormattableString.Invariant($"{first:0000}{second:0000}")
```

What do you think about having it suggest `FormattableString.Invariant` in the first place when it does not change meaning and it brings a simplification opportunity? For example:

```cs
$"{first.ToString("0000", CultureInfo.InvariantCulture)}{second.ToString("0000", CultureInfo.InvariantCulture)}"

// 💡 Simplify interpolation:
FormattableString.Invariant($"{first:0000}{second:0000}")
```

Should this be something that "Convert concatenation to interpolation" also does for you, in case you didn't realize that FormattableString.Invariant existed or could help?

```cs
first.ToString("0000", CultureInfo.InvariantCulture) + second.ToString("0000", CultureInfo.InvariantCulture)

// 💡 Convert concatenation to interpolation:
FormattableString.Invariant($"{first:0000}{second:0000}")
```

### Where to bail

The suggestion should not be given because it would change the meaning of `{42}` from current-culture to invariant-culture:
```cs
$"{42}{second.ToString("0000", CultureInfo.InvariantCulture)}"
```

The suggestion should not be given because it would switch to producing a string instead of a FormattableString instance:
```cs
FormattableString f = $"{42}{second.ToString("0000", CultureInfo.InvariantCulture)}"
```

So all interpolation sections must be known to be indifferent to culture (i.e. strings themselves) or must explicitly be using invariant formatting, and the interpolated string operation must be target-typed as string rather than FormattableString.

This should also bail because it would not change meaning but would still be confusing to wrap this in FormattableString.Invariant while leaving `.ToString(CultureInfo.CurrentCulture)` inside:
```cs
$"{first.ToString(CultureInfo.CurrentCulture)}{second.ToString(CultureInfo.InvariantCulture)}"
```

These mixed-culture strings seem highly unlikely to be intentional if they do occur.

**Comment by jasonmalinowski** [comment]
This was discussed in the design meeting and the proposal generally supported. There was one specific question about how to define this phrase:

> must be known to be indifferent to culture (i.e. strings themselves)

For example, consider the (probably common) case of a type that implements ToString, and in that calls ToString on some internal fields. It doesn't implement IFormattable or take any overloads that take a culture. Do we call this culture sensitive (since the internal fields should be sensitive?) or it's insensitive (since it has no explicit way to be sensitive to a specific culture?) @sharwell offered to write up some precise definitions for the behavior options, and we decided to let the implementer of this decide based on either what was easier or what ultimately made sense.

**Comment by jnm2** [comment]
> Do we call this culture sensitive (since the internal fields should be sensitive?) or it's insensitive (since it has no explicit way to be sensitive to a specific culture?)

Good point. I meant this to be defined based on the effect `FormattableString.Invariant` has on it, which is to say that `myClass` (if it doesn't implement IFormattable) and `myClass.ToString()` would be allowed to end up inside a `FormattableString.Invariant` suggestion. The concern was that the code should not change meaning.

It would technically be fine to allow `x.ToString(CultureInfo.CurrentCulture)` to end up inside a `FormattableString.Invariant` suggestion, but the concern in this case was reader confusion rather than meaning change.

**Comment by sharwell** [comment]
There are multiple potential ways to define "known to be indifferent to culture":

1. The type implements `IFormattable`, but is known to ignore the culture parameter (if any). `System.String` is an example of this.
2. The type does not implement `IFormattable`, and therefore doesn't not provide culture-sensitive `ToString` operations.
    * If we use this definition, we have the option of further restricting it to only consider sealed types
